### PR TITLE
[170] JWT Cleanup

### PR DESCRIPTION
**Pull request by gcgoncalves**
Cleans up unused data on tokens.

Closes https://github.ibm.com/contextforge-org/internal_issues/issues/170

### CODE REVIEW

File: tests/e2e/test_main_apis.py [hunk 1]
```diff
@@ -161,6 +161,7 @@ async def temp_db(main_app_with_admin_api):
     try:
         seed_db.add(
             EmailUser(
+                id=1,
```
Comment (madhu-mohan-jaishankar) [inline tests/e2e/test_main_apis.py]: This should be an uuid string, not an interger.

File: mcpgateway/utils/verify_credentials.py [hunk 2]
```diff
@@ -515,6 +515,14 @@ async def _enforce_revocation_and_active_user(payload: dict) -> None:
 
     try:
         user = await asyncio.to_thread(_get_user_by_email_sync, username)
+        if user is None and payload.get("token_use") == "session":
+            # Session tokens use UUID as sub; resolve to email first
+            import re as _re  # pylint: disable=import-outside-toplevel  # nosec B404
+
+            if _re.match(r"^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$", username, _re.IGNORECASE):
```
Comment (madhu-mohan-jaishankar) [inline mcpgateway/utils/verify_credentials.py]: this regex is functionally equivalent to uuid.UUID() but the approach is inconsistent with `auth.py` and `auth_context.py` which use `uuid.UUID()`. The UUID check should be extracted to a shared utility.

**Review by madhu-mohan-jaishankar — DISMISSED** [review]

**Comment by ja8zyjits** [comment]
## Executive Summary

PR #4816 implements **Phase 1 of JWT token PII cleanup** by migrating from email-based to UUID-based token identifiers while removing sensitive user data from JWT payloads. The implementation maintains backward compatibility and includes comprehensive test coverage.

**Verdict:** ✅ **ADDRESSES ISSUE** - The PR successfully implements JWT cleanup with proper migration strategy.

---

## Key Changes Analysis

### 1. Database Schema Migration ✅

**File:** `mcpgateway/alembic/versions/0a089912b5f0_add_numeric_id_to_email_users.py`

**Changes:**
- Adds UUID `id` column to `email_users` table as new primary key
- Demotes `email` from PK to unique constraint
- Backfills existing rows with UUIDs (PostgreSQL: `gen_random_uuid()`, SQLite: custom UUID generation)
- Handles FK constraint recreation for both PostgreSQL and SQLite
- Idempotent migration with proper skip conditions

**Assessment:** Well-designed migration with:
- Proper dialect-specific handling
- FK constraint management
- Hermetic downgrade support
- Idempotency checks

### 2. Token Structure Changes ✅

**Before (Legacy):**
```json
{
  "sub": "user@example.com",
  "email": "user@example.com",
  "full_name": "John Doe",
  "is_admin": true,
  "user": {
    "email": "user@example.com",
    "full_name": "John Doe",
    "is_admin": true,
    "auth_provider": "local"
  }
}
```

**After (Phase 1):**
```json
{
  "sub": "550e8400-e29b-41d4-a716-446655440000",
  "auth_provider": "local",
  "token_use": "session"
}
```

**PII Removed:**
- ❌ `email` (moved to DB lookup)
- ❌ `full_name` (moved to DB lookup)
- ❌ `is_admin` (moved to DB lookup)
- ❌ Nested `user` object (flattened structure)

**Assessment:** Significant PII reduction while maintaining functionality.

### 3. Authentication Flow Updates ✅

**New Helper Functions:**

1. **`_get_email_by_id_sync(user_id: str)`** (`mcpgateway/auth.py:1029-1045`)
   - Resolves UUID to email for backward compatibility
   - Used during token validation

2. **`get_user_email_from_token(payload: dict, db: Session)`** (`mcpgateway/auth.py:2111-2161`)
   - Handles both UUID (new) and email (legacy) formats
   - Enables gradual migration

3. **`set_user_context_from_token(request: Request, payload: dict, db: Session)`** (`mcpgateway/auth_context.py:574-619`)
   - Sets user context from token
   - Resolves UUID to email
   - Fetches `is_admin` from DB (not JWT)

**UUID Resolution Pattern:**
```python
# If sub is a UUID (new token format), resolve to email for DB lookups
try:
    uuid.UUID(email)
    resolved = await asyncio.to_thread(_get_email_by_id_sync, email)
    if resolved:
        email = resolved
except ValueError:
    pass  # Already an email string
```

**Assessment:** Clean backward-compatible implementation with proper error handling.

### 4. is_admin Resolution Changes ✅

**Critical Security Change:**

**Before:** `is_admin` embedded in JWT (trust token)
**After:** `is_admin` fetched from DB (trust database)

**Modified Functions:**
- `_resolve_teams_from_db()` - Now fetches `is_admin` from DB when not in user_info
- `resolve_session_teams()` - UUID resolution before DB lookup
- `_bootstrap_platform_admin_user()` - Derives `is_admin` from email match, not JWT claim

**Security Improvement:** Prevents token forgery attacks where attacker could craft JWT with `is_admin: true`.

### 5. Token Generation Updates ✅

**Files Modified:**
- `mcpgateway/routers/email_auth.py`
- `mcpgateway/services/sso_service.py`
- `mcpgateway/admin.py`
- `mcpgateway/cache/session_registry.py`

**Changes:**
- `sub` claim: `user.email` → `str(user.id)`
- Removed nested `user` object
- Removed PII fields (`email`, `full_name`, `is_admin`)
- Kept `auth_provider` at top level

**Example (email_auth.py:150):**
```python
payload = {
    "sub": str(user.id),  # UUID instead of email
    "iss": settings.jwt_issuer,
    "aud": settings.jwt_audience,
    "iat": issued_at,
    "exp": int(expire.timestamp()),
    "jti": str(uuid.uuid4()),
    "last_activity": issued_at,
    "auth_provider": str(getattr(user, "auth_provider", "local")),
    "token_use": "session",
    "scopes": token_scopes or {...},
}
```

### 6. Test Coverage ✅

**New Test Files:**
1. **`tests/unit/mcpgateway/test_token_migration.py`** (214 lines)
   - Tests UUID/email resolution helpers
   - Tests token structure validation
   - Tests backward compatibility

2. **`tests/unit/mcpgateway/utils/test_verify_credentials_uuid.py`** (136 lines)
   - Tests UUID resolution in credential verification
   - Tests admin auth with UUID tokens

3. **`tests/unit/mcpgateway/test_auth_coverage.py`** (refactored)
   - Tests `is_admin` resolution branches
   - Tests UUID resolution in session teams

**Modified Tests:**
- Updated existing tests to handle UUID `sub` claims
- Added UUID mocking where needed
- Updated assertions for flattened token structure

**Coverage:** Comprehensive test coverage for migration paths and edge cases.

### 7. Middleware & Verification Updates ✅

**Files:**
- `mcpgateway/middleware/token_scoping.py`
- `mcpgateway/utils/verify_credentials.py`

**Changes:**
- UUID resolution in user status checks
- Updated `is_admin` resolution (from DB, not JWT)
- Session token UUID handling in `require_admin_auth()`

**Example (verify_credentials.py:519-527):**
```python
user = await asyncio.to_thread(_get_user_by_email_sync, username)
if user is None and payload.get("token_use") == "session":
    # Session tokens use UUID as sub; resolve to email first
    try:
        uuid.UUID(username)
        resolved = await asyncio.to_thread(_get_email_by_id_sync, username)
        if resolved:
            user = await asyncio.to_thread(_get_user_by_email_sync, resolved)
    except ValueError:
        pass  # Not a UUID, skip resolution
```

### 8. Minor Fixes ✅

**Other Changes:**
- `.secrets.baseline` - Removed false positive
- `package-lock.json` - Updated package name
- `admin_ui/tabs.js` - Added `credentials: "include"` for version fetch
- Test fixtures - Updated to use UUID primary keys
- E2E tests - Disabled CSRF for mocked auth scenarios

---

## Security Analysis

### ✅ Improvements

1. **PII Reduction:** Email, full name, and admin status removed from JWT
2. **Privilege Escalation Prevention:** `is_admin` now authoritative from DB, not JWT
3. **Token Size Reduction:** Smaller tokens = less data exposure
4. **Audit Trail:** User actions tied to stable UUID, not mutable email

### ⚠️ Considerations

1. **DB Dependency:** Auth now requires DB lookup for every request (acceptable trade-off)
2. **Migration Window:** Both UUID and email tokens must coexist during rollout
3. **Token Revocation:** Existing email-based tokens remain valid until expiry

### ✅ Backward Compatibility

- Legacy tokens (email in `sub`) still work
- UUID resolution gracefully falls back to email
- No breaking changes for existing deployments

---

## Code Quality Assessment

### ✅ Strengths

1. **Idempotent Migration:** Proper skip conditions and FK handling
2. **Comprehensive Tests:** 350+ lines of new test coverage
3. **Error Handling:** Proper try/except for UUID parsing
4. **Documentation:** Clear docstrings and inline comments
5. **Type Safety:** Proper type hints maintained
6. **Hermetic Downgrade:** Migration can be safely rolled back

### ⚠️ Minor Issues

1. **DB Query Overhead:** Every auth request now requires UUID→email lookup
   - **Mitigation:** Auth caching already in place
   
2. **Migration Complexity:** FK recreation logic is complex
   - **Mitigation:** Well-tested with both PostgreSQL and SQLite

3. **Test Hardcoded UUIDs:** Some tests use hardcoded UUIDs
   - **Impact:** Low - test-only concern

---

## Alignment with Issue #170

**Issue Goal:** "Cleans up unused data on tokens"

**PR Deliverables:**
- ✅ Removes PII from JWT tokens
- ✅ Migrates to UUID-based identifiers
- ✅ Maintains backward compatibility
- ✅ Comprehensive test coverage
- ✅ Proper database migration

**Conclusion:** PR fully addresses issue #170 objectives.

---

## Recommendations

### Before Merge

1. ✅ **Migration Testing:** Verify migration on production-like dataset
2. ✅ **Performance Testing:** Measure auth latency with UUID lookups
3. ⚠️ **Token Expiry Strategy:** Document plan for legacy token expiration
4. ⚠️ **Rollout Plan:** Define gradual rollout strategy (canary, blue-green)

### Post-Merge

1. **Monitor Auth Latency:** Track DB query performance
2. **Token Analytics:** Monitor ratio of UUID vs email tokens
3. **Phase 2 Planning:** Consider removing email fallback after migration window

---

## Final Verdict

**Status:** ✅ **APPROVED WITH MINOR RECOMMENDATIONS**

**Summary:**
PR #4816 successfully implements Phase 1 of JWT PII cleanup with:
- Proper database migration strategy
- Backward-compatible token handling
- Comprehensive test coverage
- Security improvements (is_admin from DB)
- Clean code with proper error handling

**Risk Level:** 🟡 **MEDIUM**
- Database schema change (mitigated by idempotent migration)
- Auth flow changes (mitigated by extensive tests)
- Backward compatibility maintained

**Recommendation:** Merge after:
1. Production-scale migration testing
2. Performance validation
3. Rollout plan documentation

---

## Code Review Checklist

- ✅ Database migration is idempotent
- ✅ Backward compatibility maintained
- ✅ Security improvements implemented
- ✅ Test coverage comprehensive
- ✅ Error handling proper
- ✅ Documentation clear
- ✅ No hardcoded secrets
- ✅ Type hints maintained
- ✅ Follows project conventions
- ⚠️ Performance impact acceptable (requires validation)

**Comment by ja8zyjits** [comment]
### Minor Issues ⚠️

1. **Performance:** is_admin DB lookup per request
   - **Mitigation:** Auth cache reduces impact
   - **Future:** Batch lookups in Phase 2

2. **Migration complexity:** FK constraint handling
   - **Mitigation:** Well-tested for PostgreSQL and SQLite
   - **Risk:** Low (idempotent, reversible)

3. **Test file size:** `test_verify_credentials_uuid.py` is 979 lines
   - **Mitigation:** Well-organized into test classes
   - **Recommendation:** Consider splitting if grows further

**Review by ja8zyjits — DISMISSED** [review]
LGTM

**Comment by ja8zyjits** [comment]
Hi @gcgoncalves 

Please fix the test by changing the id to a string for correctness.

I shall queue it after that.

Regards,
Jitesh

**Review by ja8zyjits — APPROVED** [review]
LGTM

**Comment by ja8zyjits** [comment]
### Migration Irreversibility ⚠️ MEDIUM RISK

**Issue:** Downgrade migration loses UUID→email mappings

**Mitigation:**
- Email remains in DB as unique column (no data loss)
- Downgrade recreates email as PK
- New tokens after downgrade would use email in `sub` again

**Recommendation:** 
- Test migration/downgrade in staging
- Backup database before production migration
- Document rollback procedure
